Runtime failed due to below error, which is highlighted in master as well.

```
pkg/maps/lbmap/maglev_test.go:55:3: unknown field 'MaxEntries' in struct literal of type InitParams
```

--- a/operator/pkg/ingress/envoy_config.go
+++ b/operator/pkg/ingress/envoy_config.go
@@ -8,6 +8,7 @@ import (
 	"fmt"
 	"net"
 	"sort"
+	"strings"
 
 	envoy_config_cluster_v3 "github.com/cilium/proxy/go/envoy/config/cluster/v3"
 	envoy_config_core_v3 "github.com/cilium/proxy/go/envoy/config/core/v3"
@@ -35,7 +36,10 @@ import (
 	slim_networkingv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/networking/v1"
 )
 
-const wildCard = "*"
+const (
+	wildCard       = "*"
+	envoyAuthority = ":authority"
+)
 
 type envoyConfigManager struct {
 	informer   cache.Controller
@@ -403,7 +407,8 @@ func toAny(message proto.Message) *anypb.Any {
 	return a
 }
 
-func getRouteMatch(ingressPath slim_networkingv1.HTTPIngressPath) *envoy_config_route_v3.RouteMatch {
+func getRouteMatch(host string, ingressPath slim_networkingv1.HTTPIngressPath) *envoy_config_route_v3.RouteMatch {
+	headerMatchers := getHeaderMatchers(host)
 	if ingressPath.PathType == nil || *ingressPath.PathType == slim_networkingv1.PathTypeImplementationSpecific ||
 		*ingressPath.PathType == slim_networkingv1.PathTypePrefix {
 		return &envoy_config_route_v3.RouteMatch{
@@ -413,6 +418,7 @@ func getRouteMatch(ingressPath slim_networkingv1.HTTPIngressPath) *envoy_config_
 					Regex:      getMatchingPrefixRegex(ingressPath.Path),
 				},
 			},
+			Headers: headerMatchers,
 		}
 	}
 	if *ingressPath.PathType == slim_networkingv1.PathTypeExact {
@@ -420,16 +426,40 @@ func getRouteMatch(ingressPath slim_networkingv1.HTTPIngressPath) *envoy_config_
 			PathSpecifier: &envoy_config_route_v3.RouteMatch_Path{
 				Path: ingressPath.Path,
 			},
+			Headers: headerMatchers,
 		}
 	}
 	return nil
+}
+
+func getHeaderMatchers(host string) []*envoy_config_route_v3.HeaderMatcher {
+	if len(host) == 0 || host == wildCard || !strings.Contains(host, wildCard) {
+		return nil
+	}
+	// Make sure that wildcard character only match one single dns domain.
+	// For example, if host is *.foo.com, baz.bar.foo.com should not match
+	return []*envoy_config_route_v3.HeaderMatcher{
+		{
+			Name: envoyAuthority,
+			HeaderMatchSpecifier: &envoy_config_route_v3.HeaderMatcher_StringMatch{
+				StringMatch: &envoy_type_matcher_v3.StringMatcher{
+					MatchPattern: &envoy_type_matcher_v3.StringMatcher_SafeRegex{
+						SafeRegex: &envoy_type_matcher_v3.RegexMatcher{
+							EngineType: &envoy_type_matcher_v3.RegexMatcher_GoogleRe2{},
+							Regex:      getMatchingHeaderRegex(host),
+						},
+					},
+				},
+			},
+		},
+	}
 }
 
 func getVirtualHost(ingress *slim_networkingv1.Ingress, rule slim_networkingv1.IngressRule) *envoy_config_route_v3.VirtualHost {
 	routes := make(SortableRoute, 0, len(rule.HTTP.Paths))
 	for _, path := range rule.HTTP.Paths {
 		route := envoy_config_route_v3.Route{
-			Match: getRouteMatch(path),
+			Match: getRouteMatch(rule.Host, path),
 			Action: &envoy_config_route_v3.Route_Route{
 				Route: &envoy_config_route_v3.RouteAction{
 					ClusterSpecifier: &envoy_config_route_v3.RouteAction_Cluster{

--- a/operator/pkg/ingress/envoy_config.go
+++ b/operator/pkg/ingress/envoy_config.go
@@ -426,7 +426,7 @@ func getRouteMatch(ingressPath slim_networkingv1.HTTPIngressPath) *envoy_config_
 }
 
 func getVirtualHost(ingress *slim_networkingv1.Ingress, rule slim_networkingv1.IngressRule) *envoy_config_route_v3.VirtualHost {
-	var routes []*envoy_config_route_v3.Route
+	routes := make(SortableRoute, 0, len(rule.HTTP.Paths))
 	for _, path := range rule.HTTP.Paths {
 		route := envoy_config_route_v3.Route{
 			Match: getRouteMatch(path),
@@ -441,6 +441,11 @@ func getVirtualHost(ingress *slim_networkingv1.Ingress, rule slim_networkingv1.I
 
 		routes = append(routes, &route)
 	}
+
+	// This is to make sure that the Exact match is always having higher priority.
+	// Each route entry in the virtual host is checked, in order. If there is a match, the route is used and no further route checks are made.
+	// Related docs https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/route_matching
+	sort.Sort(routes)
 
 	domains := []string{wildCard}
 	if rule.Host != "" {

--- a/operator/pkg/ingress/envoy_config.go
+++ b/operator/pkg/ingress/envoy_config.go
@@ -16,6 +16,7 @@ import (
 	envoy_extensions_filters_network_http_connection_manager_v3 "github.com/cilium/proxy/go/envoy/extensions/filters/network/http_connection_manager/v3"
 	envoy_extensions_transport_sockets_tls_v3 "github.com/cilium/proxy/go/envoy/extensions/transport_sockets/tls/v3"
 	envoy_config_upstream "github.com/cilium/proxy/go/envoy/extensions/upstreams/http/v3"
+	envoy_type_matcher_v3 "github.com/cilium/proxy/go/envoy/type/matcher/v3"
 	"google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/types/known/anypb"
 	"google.golang.org/protobuf/types/known/durationpb"
@@ -406,8 +407,11 @@ func getRouteMatch(ingressPath slim_networkingv1.HTTPIngressPath) *envoy_config_
 	if ingressPath.PathType == nil || *ingressPath.PathType == slim_networkingv1.PathTypeImplementationSpecific ||
 		*ingressPath.PathType == slim_networkingv1.PathTypePrefix {
 		return &envoy_config_route_v3.RouteMatch{
-			PathSpecifier: &envoy_config_route_v3.RouteMatch_Prefix{
-				Prefix: ingressPath.Path,
+			PathSpecifier: &envoy_config_route_v3.RouteMatch_SafeRegex{
+				SafeRegex: &envoy_type_matcher_v3.RegexMatcher{
+					EngineType: &envoy_type_matcher_v3.RegexMatcher_GoogleRe2{},
+					Regex:      getMatchingPrefixRegex(ingressPath.Path),
+				},
 			},
 		}
 	}

--- a/operator/pkg/ingress/envoy_config_test.go
+++ b/operator/pkg/ingress/envoy_config_test.go
@@ -149,7 +149,7 @@ func Test_getRouteConfigurationResource(t *testing.T) {
 	require.Len(t, routeConfig.VirtualHosts[1].Routes, 1)
 
 	require.Equal(t, "/dummy-path", routeConfig.VirtualHosts[0].Routes[0].Match.GetPath())
-	require.Equal(t, "/another-dummy-path", routeConfig.VirtualHosts[0].Routes[1].Match.GetPrefix())
+	require.Equal(t, "/another-dummy-path(/.*)?$", routeConfig.VirtualHosts[0].Routes[1].Match.GetSafeRegex().GetRegex())
 	require.Equal(t, "/", routeConfig.VirtualHosts[1].Routes[0].Match.GetPrefix())
 
 	clusters := []string{routeConfig.VirtualHosts[0].Routes[0].GetRoute().GetCluster(), routeConfig.VirtualHosts[0].Routes[1].GetRoute().GetCluster()}

--- a/operator/pkg/ingress/envoy_config_test.go
+++ b/operator/pkg/ingress/envoy_config_test.go
@@ -143,6 +143,7 @@ func Test_getRouteConfigurationResource(t *testing.T) {
 	require.Equal(t, "*", routeConfig.VirtualHosts[0].Name)
 	require.Equal(t, []string{"*"}, routeConfig.VirtualHosts[0].Domains)
 	require.Len(t, routeConfig.VirtualHosts[0].Routes, 2)
+	require.Len(t, routeConfig.VirtualHosts[0].Routes[0].Match.GetHeaders(), 0)
 
 	require.Equal(t, "default-backend", routeConfig.VirtualHosts[1].Name)
 	require.Equal(t, []string{"*"}, routeConfig.VirtualHosts[1].Domains)
@@ -151,6 +152,7 @@ func Test_getRouteConfigurationResource(t *testing.T) {
 	require.Equal(t, "/dummy-path", routeConfig.VirtualHosts[0].Routes[0].Match.GetPath())
 	require.Equal(t, "/another-dummy-path(/.*)?$", routeConfig.VirtualHosts[0].Routes[1].Match.GetSafeRegex().GetRegex())
 	require.Equal(t, "/", routeConfig.VirtualHosts[1].Routes[0].Match.GetPrefix())
+	require.Len(t, routeConfig.VirtualHosts[1].Routes[0].Match.GetHeaders(), 0)
 
 	clusters := []string{routeConfig.VirtualHosts[0].Routes[0].GetRoute().GetCluster(), routeConfig.VirtualHosts[0].Routes[1].GetRoute().GetCluster()}
 	require.Contains(t, clusters, "dummy-namespace/dummy-backend")

--- a/operator/pkg/ingress/regex_helper.go
+++ b/operator/pkg/ingress/regex_helper.go
@@ -8,7 +8,13 @@ import (
 	"strings"
 )
 
-const slash = "/"
+const (
+	slash       = "/"
+	dot         = "."
+	starDot     = "*."
+	dotRegex    = "[.]"
+	notDotRegex = "[^.]"
+)
 
 // getMatchingPrefixRegex returns safe regex used by envoy to match the prefix.
 // By default, prefix matching in envoy will not reject /foobar if the original path is only /foo. Hence, conversion
@@ -24,4 +30,13 @@ func getMatchingPrefixRegex(path string) string {
 		removedTrailingSlash = removedTrailingSlash[:len(removedTrailingSlash)-1]
 	}
 	return fmt.Sprintf("%s(/.*)?$", removedTrailingSlash)
+}
+
+// getMatchingHeaderRegex is to make sure that one and only one single subdomain is matched e.g.
+// For example, *.foo.com should only match bar.foo.com but not baz.bar.foo.com
+func getMatchingHeaderRegex(host string) string {
+	if strings.HasPrefix(host, starDot) {
+		return fmt.Sprintf("^%s+%s%s$", notDotRegex, dotRegex, strings.ReplaceAll(host[2:], dot, dotRegex))
+	}
+	return fmt.Sprintf("^%s$", strings.ReplaceAll(host, dot, dotRegex))
 }

--- a/operator/pkg/ingress/regex_helper.go
+++ b/operator/pkg/ingress/regex_helper.go
@@ -1,0 +1,27 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ingress
+
+import (
+	"fmt"
+	"strings"
+)
+
+const slash = "/"
+
+// getMatchingPrefixRegex returns safe regex used by envoy to match the prefix.
+// By default, prefix matching in envoy will not reject /foobar if the original path is only /foo. Hence, conversion
+// with safe regex is required.
+//
+// If the original path is /foo, the returned regex will be /foo(/.*)?$
+// - /foo -> matched
+// - /foo/ -> matched
+// - /foobar -> not matched
+func getMatchingPrefixRegex(path string) string {
+	removedTrailingSlash := path
+	if strings.HasSuffix(path, slash) {
+		removedTrailingSlash = removedTrailingSlash[:len(removedTrailingSlash)-1]
+	}
+	return fmt.Sprintf("%s(/.*)?$", removedTrailingSlash)
+}

--- a/operator/pkg/ingress/regex_helper_test.go
+++ b/operator/pkg/ingress/regex_helper_test.go
@@ -1,0 +1,48 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+//go:build !privileged_tests
+
+package ingress
+
+import (
+	"regexp"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_getMatchingPrefixRegex(t *testing.T) {
+	t.Run("simple slash path", func(t *testing.T) {
+		rx, err := regexp.Compile(getMatchingPrefixRegex("/"))
+		assert.NoError(t, err)
+
+		assert.True(t, rx.MatchString("/"))
+		assert.True(t, rx.MatchString("/foo"))
+		assert.True(t, rx.MatchString("/foo/bar"))
+	})
+
+	t.Run("simple /foo path", func(t *testing.T) {
+		rx, err := regexp.Compile(getMatchingPrefixRegex("/foo"))
+		assert.NoError(t, err)
+
+		assert.False(t, rx.MatchString("/"))
+		assert.False(t, rx.MatchString("/foobar"))
+
+		assert.True(t, rx.MatchString("/foo"))
+		assert.True(t, rx.MatchString("/foo/"))
+		assert.True(t, rx.MatchString("/foo/bar"))
+	})
+
+	t.Run("path with trailing slash /foo/ path", func(t *testing.T) {
+		rx, err := regexp.Compile(getMatchingPrefixRegex("/foo/"))
+		assert.NoError(t, err)
+
+		assert.False(t, rx.MatchString("/"))
+		assert.False(t, rx.MatchString("/foobar"))
+
+		assert.True(t, rx.MatchString("/foo"))
+		assert.True(t, rx.MatchString("/foo/"))
+		assert.True(t, rx.MatchString("/foo/bar"))
+	})
+}

--- a/operator/pkg/ingress/regex_helper_test.go
+++ b/operator/pkg/ingress/regex_helper_test.go
@@ -46,3 +46,33 @@ func Test_getMatchingPrefixRegex(t *testing.T) {
 		assert.True(t, rx.MatchString("/foo/bar"))
 	})
 }
+
+func Test_getMatchingHeaderRegex(t *testing.T) {
+	t.Run("starting with *.", func(t *testing.T) {
+		result := getMatchingHeaderRegex("*.foo.com")
+		rx, err := regexp.Compile(result)
+		assert.NoError(t, err)
+
+		// Match entire host
+		assert.True(t, rx.MatchString("bar.foo.com"))
+
+		// Not match
+		assert.False(t, rx.MatchString("foo.com"))
+		assert.False(t, rx.MatchString("baz.bar.foo.com"))
+		assert.False(t, rx.MatchString("some.random.host.com"))
+	})
+
+	t.Run("not starting with *.", func(t *testing.T) {
+		result := getMatchingHeaderRegex("foo.com")
+		rx, err := regexp.Compile(result)
+		assert.NoError(t, err)
+
+		// Match entire host
+		assert.True(t, rx.MatchString("foo.com"))
+
+		// Not match
+		assert.False(t, rx.MatchString("bar.foo.com"))
+		assert.False(t, rx.MatchString("abc.bar.foo.com"))
+		assert.False(t, rx.MatchString("some.random.host.com"))
+	})
+}

--- a/operator/pkg/ingress/types.go
+++ b/operator/pkg/ingress/types.go
@@ -1,0 +1,39 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ingress
+
+import envoy_config_route_v3 "github.com/cilium/proxy/go/envoy/config/route/v3"
+
+// SortableRoute is a slice of envoy Route, which can be sorted based on matching order as per Ingress requirement.
+//
+// - Exact Match must have the highest priority
+// - If multiple prefix matches are satisfied, the longest path is having higher priority
+//
+// As Envoy route matching logic is done sequentially, we need to enforce such sorting order.
+type SortableRoute []*envoy_config_route_v3.Route
+
+func (s SortableRoute) Len() int {
+	return len(s)
+}
+
+func (s SortableRoute) Less(i, j int) bool {
+	// Make sure Exact Match always comes first
+	isExactMatch1 := len(s[i].Match.GetPath()) != 0
+	isExactMatch2 := len(s[j].Match.GetPath()) != 0
+	if isExactMatch1 {
+		return true
+	}
+	if isExactMatch2 {
+		return false
+	}
+
+	// Make sure longest Prefix match always comes first
+	regexMatch1 := len(s[i].Match.GetSafeRegex().String())
+	regexMatch2 := len(s[j].Match.GetSafeRegex().String())
+	return regexMatch1 > regexMatch2
+}
+
+func (s SortableRoute) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}

--- a/operator/pkg/ingress/types_test.go
+++ b/operator/pkg/ingress/types_test.go
@@ -1,0 +1,68 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+//go:build !privileged_tests
+
+package ingress
+
+import (
+	"sort"
+	"testing"
+
+	envoy_config_route_v3 "github.com/cilium/proxy/go/envoy/config/route/v3"
+	envoy_type_matcher_v3 "github.com/cilium/proxy/go/envoy/type/matcher/v3"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSortableRoute(t *testing.T) {
+	arr := SortableRoute{
+		{
+			Name: "exact match 1",
+			Match: &envoy_config_route_v3.RouteMatch{
+				PathSpecifier: &envoy_config_route_v3.RouteMatch_Path{
+					Path: "/exact/match",
+				},
+			},
+		},
+		{
+			Name: "another exact match",
+			Match: &envoy_config_route_v3.RouteMatch{
+				PathSpecifier: &envoy_config_route_v3.RouteMatch_Path{
+					Path: "/exact/match/another",
+				},
+			},
+		},
+		{
+			Name: "prefix match",
+			Match: &envoy_config_route_v3.RouteMatch{
+				PathSpecifier: &envoy_config_route_v3.RouteMatch_SafeRegex{
+					SafeRegex: &envoy_type_matcher_v3.RegexMatcher{
+						EngineType: &envoy_type_matcher_v3.RegexMatcher_GoogleRe2{},
+						Regex:      "/prefix/match",
+					},
+				},
+			},
+		},
+		{
+			Name: "another prefix match",
+			Match: &envoy_config_route_v3.RouteMatch{
+				PathSpecifier: &envoy_config_route_v3.RouteMatch_SafeRegex{
+					SafeRegex: &envoy_type_matcher_v3.RegexMatcher{
+						EngineType: &envoy_type_matcher_v3.RegexMatcher_GoogleRe2{},
+						Regex:      "/prefix/match/another",
+					},
+				},
+			},
+		},
+	}
+
+	sort.Sort(arr)
+
+	// Exact match comes first in any order
+	assert.True(t, len(arr[0].Match.GetPath()) != 0)
+	assert.True(t, len(arr[1].Match.GetPath()) != 0)
+
+	// Prefix match with longer path comes first
+	assert.Equal(t, "/prefix/match/another", arr[2].Match.GetSafeRegex().GetRegex())
+	assert.Equal(t, "/prefix/match", arr[3].Match.GetSafeRegex().GetRegex())
+}

